style-guide.md

```diff
@@ -0,0 +1,65 @@
+    - In most cases, just passing the parameter to the underlying tool should produce a satisfactory error, but this must be checked for each task
+- All tasks must have configurable memory and disk space allocations
+  - TODO: only enforceable if we require specific var naming. Do we want to enforce specific variable names? Keep this in the guide?
+  - see the various tasks in the template directory for possible ways to allocate resources
+    - Contributors can mix and match the available templates, copy and pasting subsections as appropriate
+    - It is allowed to have one resource allocated dynamically, and another allocated statically in the same task.
+    - It is *not* allowed to have a resource which can be allocated *either* statically or dynamically.
+      - This is technically feasible, but is too complicated for maintenance and end-users.
+      - e.g. `memory_gb` and `modify_memory_gb` cannot be present in the same task.
+- All tasks and workflows should have a `max_retries` input.
+  - TODO: again, only enforeceable with specific var names. Not sure if we want to get into that business. Keep it in the guide?
+  - This should be defaulted to `1` for nearly all tasks
+  - Some tasks are particularly error prone and can have a higher default `max_retries`
+  - **rule specific to workflows:** `max_retries` should be an optional `Int?`
+    - This allows each task to have it's own specific default `max_retries`
+      - If a user does not supply `max_retries`, those task level defaults will get used
+      - If a user does supply `max_retries`, it should override the default for *every* task called
+- multi-core tasks should *always* follow the conventions laid out in the `detect_nproc_task` example (see `template/task-templates.wdl`)
+  - this is catering to cloud users, who may be allocated a machine with more cores than are specified by the `ncpu` parameter
+  - TODO: this would be tough to validate with a linter. It is definitely a best practice, but IMO doesn't belong in the guide
+- `command` blocks should be wrapped with arrows (`<<< >>>`) instead of brackets (`{ }`)
+  - Certain Bash constructions cause problems with the bracket notation
+- output file names should *always* be determined with either the `outfile_name` parameter or the `prefix` parameter.
+  - `outfile_name` should be preferred if no downstream tasks/tools rely on the file name/extension
+  - tasks with multiple outputs should always use the `prefix` convention
+  - TODO: IMO another "best practice" that shouldn't be enforced by a linter. Remove from guide?
+- output variable names should be short but descriptive
+  - TODO: This I'm more inclined to enforce with the linter. I think disallowing specific output names is better than forcing users into using specific ones.
+  - Disallowed output names: `out*`, any name less than 3 characters
+    - TODO: is this sufficient?
```

Comment:
I think this is reasonable to enforce. Getting the pattern right is going to be challenging. 3+ characters is probably a good start. Can you think of any descriptive name that would start with `out`, maybe something with `outer` or `outside` though I lack a good example?

Reply:
`outer*` and `outside*` are good exceptions. Maybe also `outsource*`, `outward*`, `outrange*`, `outbreak*`, `outskirt*`, `outlie*`, and `outlay*`?

I just googled "words that start with out" and went through [this list](https://www.thefreedictionary.com/words-that-start-with-out) (which took longer than I'd like to admit. More words start with "out" than I realized lol). But I kept an eye out for anything that might make sense for this context, and I think we'd be fine banning anything not in the above exceptions.

Reply:
Note that I'm skirting the issue of enforcing a case system. WDL itself is inconsistent (e.g. `maxRetries` and `parameter_meta` both exist in the spec). I think:
1) it would be difficult to enforce sanely
2) not really our job
3) the language itself is inconsistent, so what's the point

I have a strong preference for snake case, but I'm not ready to enforce snake case on other people's code (now that I'm writing this comment though, I will add it to our "best practices" guide).

The regexs in the previous comment also allows a writer to get past the filter by ALL CAPSing their variable names (or at least the `IN`/`OUT` prefixes). Which... I don't really like, but I'm hesitant to cast a wider net. If someone is that insistent on writing ugly code, I don't think any tool we implement will stop them. They probably aren't our intended user 🤷